<a href="https://colab.research.google.com/github/Gaurav-Yaduvanshi/La-Liga-cup-Football-tournament-Analysis/blob/main/La_Liga_cup_(Football_tournament)_.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import scipy.stats as stats

In [71]:
path = '/content/drive/MyDrive/Colab Notebooks/great learning laliga project assignment/'
df = pd.read_csv(path + 'Laliga.csv',header=1)

In [86]:
# Number of records(Rows) and features(Columns)
df.shape

(61, 20)

In [73]:
# top 5 
df.head()

Pos             Team  Seasons Points GamesPlayed GamesWon GamesDrawn  \
0    1      Real Madrid       86   4385        2762     1647        552   
1    2        Barcelona       86   4262        2762     1581        573   
2    3  Atletico Madrid       80   3442        2614     1241        598   
3    4         Valencia       82   3386        2664     1187        616   
4    5  Athletic Bilbao       86   3368        2762     1209        633   

  GamesLost GoalsFor GoalsAgainst Champion Runner-up Third Fourth Fifth Sixth  \
0       563     5947         3140       33        23     8      8     3     4   
1       608     5900         3114       25        25    12     12     4     6   
2       775     4534         3309       10         8    16      9     7     6   
3       861     4398         3469        6         6    10     11    10     7   
4       920     4631         3700        8         7    10      5     8    10   

    T    Debut Since/LastApp  BestPosition  
0  79     1929          1929             1  
1  83     1929          1929             1  
2  56     1929       2002-03             1  
3  50  1931-32       1987-88             1  
4  49     1929          1929             1

In [74]:
# last 5
df.tail()

Pos              Team  Seasons Points GamesPlayed GamesWon GamesDrawn  \
56   57             Xerez        1     34          38        8         10   
57   58            Condal        1     22          30        7          8   
58   59   Atletico Tetuan        1     19          30        7          5   
59   60  Cultural Leonesa        1     14          30        5          4   
60   61            Girona        1      -           -        -          -   

   GamesLost GoalsFor GoalsAgainst Champion Runner-up Third Fourth Fifth  \
56        20       38           66        -         -     -      -     -   
57        15       37           57        -         -     -      -     -   
58        18       51           85        -         -     -      -     -   
59        21       34           65        -         -     -      -     -   
60         -        -            -        -         -     -      -     -   

   Sixth  T    Debut Since/LastApp  BestPosition  
56     -  -  2009-10       2009-10            20  
57     -  -  1956-57       1956-57            16  
58     -  -  1951-52       1951-52            16  
59     -  -  1955-56       1955-56            15  
60     -  -  2017-18       2017-18             9

In [76]:
# Statistical values
df.describe()

Pos    Seasons  BestPosition
count  61.000000  61.000000     61.000000
mean   31.000000  24.000000      7.081967
std    17.752934  26.827225      5.276663
min     1.000000   1.000000      1.000000
25%    16.000000   4.000000      3.000000
50%    31.000000  12.000000      6.000000
75%    46.000000  38.000000     10.000000
max    61.000000  86.000000     20.000000

In [77]:
# Summary of the data 
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 61 entries, 0 to 60
Data columns (total 20 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Pos            61 non-null     int64 
 1   Team           61 non-null     object
 2   Seasons        61 non-null     int64 
 3   Points         61 non-null     object
 4   GamesPlayed    61 non-null     object
 5   GamesWon       61 non-null     object
 6   GamesDrawn     61 non-null     object
 7   GamesLost      61 non-null     object
 8   GoalsFor       61 non-null     object
 9   GoalsAgainst   61 non-null     object
 10  Champion       61 non-null     object
 11  Runner-up      61 non-null     object
 12  Third          61 non-null     object
 13  Fourth         61 non-null     object
 14  Fifth          61 non-null     object
 15  Sixth          61 non-null     object
 16  T              61 non-null     object
 17  Debut          61 non-null     object
 18  Since/LastApp  61 non-null     o

In [78]:
# Checking Null values 
df.isna().sum()

Pos              0
Team             0
Seasons          0
Points           0
GamesPlayed      0
GamesWon         0
GamesDrawn       0
GamesLost        0
GoalsFor         0
GoalsAgainst     0
Champion         0
Runner-up        0
Third            0
Fourth           0
Fifth            0
Sixth            0
T                0
Debut            0
Since/LastApp    0
BestPosition     0
dtype: int64

In [79]:
# Using replace() to Replace dash with 0 
df.replace('-','0', inplace=True)


In [80]:
# check the last 5 
df.tail()

Pos              Team  Seasons Points GamesPlayed GamesWon GamesDrawn  \
56   57             Xerez        1     34          38        8         10   
57   58            Condal        1     22          30        7          8   
58   59   Atletico Tetuan        1     19          30        7          5   
59   60  Cultural Leonesa        1     14          30        5          4   
60   61            Girona        1      0           0        0          0   

   GamesLost GoalsFor GoalsAgainst Champion Runner-up Third Fourth Fifth  \
56        20       38           66        0         0     0      0     0   
57        15       37           57        0         0     0      0     0   
58        18       51           85        0         0     0      0     0   
59        21       34           65        0         0     0      0     0   
60         0        0            0        0         0     0      0     0   

   Sixth  T    Debut Since/LastApp  BestPosition  
56     0  0  2009-10       2009-10            20  
57     0  0  1956-57       1956-57            16  
58     0  0  1951-52       1951-52            16  
59     0  0  1955-56       1955-56            15  
60     0  0  2017-18       2017-18             9

In [81]:
# All columns
df.columns

Index(['Pos', 'Team', 'Seasons', 'Points', 'GamesPlayed', 'GamesWon',
       'GamesDrawn', 'GamesLost', 'GoalsFor', 'GoalsAgainst', 'Champion',
       'Runner-up', 'Third', 'Fourth', 'Fifth', 'Sixth', 'T', 'Debut',
       'Since/LastApp', 'BestPosition'],
      dtype='object')

# All the teams which have started playing between 1930-1980.

In [211]:
# Using astype(str) to change datatype and str[:4] for slicing four characters then to int using astype(int) and between to filter between 1930 to 1980
df_Debut_year = df[df['Debut'].astype(str).str[:4].astype(int).between(1930, 1980)]
df_Debut_year[['Team','Debut']]

Team    Debut
3              Valencia  1931-32
5               Sevilla  1934-35
8              Zaragoza  1939-40
9            Real Betis  1932-33
10  Deportivo La Coruna  1941-42
11           Celta Vigo  1939-40
12           Valladolid  1948-49
14       Sporting Gijon  1944-45
15              Osasuna  1935-36
16               Malaga  1949-50
17               Oviedo  1933-34
18             Mallorca  1960-61
19           Las Palmas  1951-52
21              Granada  1941-42
22       Rayo Vallecano  1977-78
23                Elche  1959-60
25             Hercules  1935-36
26             Tenerife  1961-62
27               Murcia  1940-41
28               Alaves  1930-31
29              Levante  1963-64
30            Salamanca  1974-75
31             Sabadell  1943-44
32                Cadiz  1977-78
34            Castellon  1941-42
37              Cordoba  1962-63
39           Recreativo  1978-79
40            Burgos CF  1971-72
41           Pontevedra  1963-64
46            Gimnastic  1947-48
49             Alcoyano  1945-46
50                 Jaen  1953-54
52           AD Almeria  1979-80
54               Lleida  1950-51
57               Condal  1956-57
58      Atletico Tetuan  1951-52
59     Cultural Leonesa  1955-56

# The list of teams which came Top 5 in terms of points.


In [100]:
df.columns

Index(['Pos', 'Team', 'Seasons', 'Points', 'GamesPlayed', 'GamesWon',
       'GamesDrawn', 'GamesLost', 'GoalsFor', 'GoalsAgainst', 'Champion',
       'Runner-up', 'Third', 'Fourth', 'Fifth', 'Sixth', 'T', 'Debut',
       'Since/LastApp', 'BestPosition'],
      dtype='object')

In [118]:
# change datatype from str to int using astype(int)

df['Points'] = df['Points'].astype(int)

In [219]:
# Sorting values for top five 5(ascending= false) 
df_top5 = df[['Team','Points']].sort_values(by = 'Points',ascending = False).head(5)
df_top5

Team  Points
0      Real Madrid    4385
1        Barcelona    4262
2  Atletico Madrid    3442
3         Valencia    3386
4  Athletic Bilbao    3368

# A function with name “Goal_diff_count” which should return all the teams with their Goal Differences

In [132]:
def Goal_diff_count(df):
  df['Goal_diff_count'] = df['GoalsFor'].astype(int) - df['GoalsAgainst'].astype(int)
  return df[['Team','Goal_diff_count']].sort_values(by = 'Goal_diff_count',ascending=False)

In [133]:
Goal_diff_count(df)

Team  Goal_diff_count
0        Real Madrid             2807
1          Barcelona             2786
2    Atletico Madrid             1225
4    Athletic Bilbao              931
3           Valencia              929
..               ...              ...
27            Murcia             -385
19        Las Palmas             -399
14    Sporting Gijon             -399
12        Valladolid             -413
13  Racing Santander             -525

[61 rows x 2 columns]

# Find the team which has maximum and minimum goal difference.

In [134]:
# Team which has the maximum goal difference
Goal_diff_count(df).head(1)

Team  Goal_diff_count
0  Real Madrid             2807

In [135]:
# Team which has the minimum goal difference
Goal_diff_count(df).tail(1)

Team  Goal_diff_count
13  Racing Santander             -525

##  Create a new column with name “Winning Percent” and append it to the data set. If there are any numerical error, replace it with 0%

In [138]:
#Percentage of Winning = (GamesWon / GamesPlayed)*100

df['Winning Percent'] = (df['GamesWon'].astype(int)/df['GamesPlayed'].astype(int))*100

df['Winning Percent'].fillna(0,inplace = True)

In [213]:
# the top 5 teams which has the highest Winning percentage

df[['Team','Goal_diff_count','Winning Percent']].sort_values(by = 'Winning Percent',ascending = False).head(5)

Team  Goal_diff_count  Winning Percent
0      Real Madrid             2807        59.630702
1        Barcelona             2786        57.241130
2  Atletico Madrid             1225        47.475134
3         Valencia              929        44.557057
4  Athletic Bilbao              931        43.772629

In [141]:
df.head()

Pos             Team  Seasons  Points GamesPlayed GamesWon GamesDrawn  \
0    1      Real Madrid       86    4385        2762     1647        552   
1    2        Barcelona       86    4262        2762     1581        573   
2    3  Atletico Madrid       80    3442        2614     1241        598   
3    4         Valencia       82    3386        2664     1187        616   
4    5  Athletic Bilbao       86    3368        2762     1209        633   

  GamesLost GoalsFor GoalsAgainst  ... Third Fourth Fifth Sixth   T    Debut  \
0       563     5947         3140  ...     8      8     3     4  79     1929   
1       608     5900         3114  ...    12     12     4     6  83     1929   
2       775     4534         3309  ...    16      9     7     6  56     1929   
3       861     4398         3469  ...    10     11    10     7  50  1931-32   
4       920     4631         3700  ...    10      5     8    10  49     1929   

  Since/LastApp BestPosition Goal_diff_count  Winning Percent  
0          1929            1            2807        59.630702  
1          1929            1            2786        57.241130  
2       2002-03            1            1225        47.475134  
3       1987-88            1             929        44.557057  
4          1929            1             931        43.772629  

[5 rows x 22 columns]

# Group teams based on their “Best position” and sum of their points for all positions

In [149]:
df.columns

Index(['Pos', 'Team', 'Seasons', 'Points', 'GamesPlayed', 'GamesWon',
       'GamesDrawn', 'GamesLost', 'GoalsFor', 'GoalsAgainst', 'Champion',
       'Runner-up', 'Third', 'Fourth', 'Fifth', 'Sixth', 'T', 'Debut',
       'Since/LastApp', 'BestPosition', 'Goal_diff_count', 'Winning Percent'],
      dtype='object')

In [208]:
# Groupby BestPositions and Points and using sum on points
df_group = df.groupby('BestPosition')['Points'].sum().sort_values(ascending = False).reset_index()

In [209]:

df_group

BestPosition  Points
0              1   27933
1              2    6904
2              4    6563
3              3    5221
4              6    2113
5              5    1884
6              7    1186
7              8    1134
8             12     511
9             10     450
10            11     445
11            17     266
12             9      96
13            16      81
14            19      81
15            14      71
16            20      34
17            15      14

# Conclusion
### 1.  There are 37 teams which started playing between 1930-1980.
### 2.  Team Real Madrid, Barcelona, Atletico Madrid, Valencia, Athletic Bilbao are the teams which came Top 5 in terms of points with 4385, 4262, 3442, 3386,  3368 points.
### 3.  Real Madrid Team has the maximum goal difference with 2807 and Racing Santander Team has the minimum goal difference with -525 goals.
### 4.  Wining Percent of Real Madrid is 59.63, Barcelona is	57.24, Girona is 0.00 and Cultural Leonesa is 16.66.
### 5.  Best Position 1  = 27933 and 2 = 6904 points.